File: 2025-01-17/Coin_Change.java
package practice;

import java.util.Arrays;

public class Coin_Change {
    public static int solve(int[] coins, int[] dp, int amount) {
        if(amount == 0) {
            return 0;
        }

        if(amount < 0) {
            return Integer.MAX_VALUE;
        }

        if(dp[amount] != -1) {
            return dp[amount];
        }

        int min = Integer.MAX_VALUE;
        for (int coin : coins) {
            int ans = solve(coins, dp, amount - coin);
            if (ans != Integer.MAX_VALUE)
                min = Math.min(min, 1 + ans);
        }

        return dp[amount] = min;
    }
    public static int coinchange2(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        Arrays.fill(dp, -1);

        int ans = solve(coins, dp, amount);
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }

    public static int coinChange(int[] coins, int amount) {
        int min = 0;
        Arrays.sort(coins);

        for(int i=coins.length-1;i>=0;i--) {
            int coin = coins[i];

            while(amount - coin >= 0) {
                amount -= coin;
                min++;
            }
        }

        return amount != 0 ? -1 : min;
    }
    public static void main(String[] args) {
        int[] coin = new int[]{186,419,83,408};
        System.out.println(coinchange2(coin, 6249));
    }
}

File: 2026-02-24/Coin_Change.py
from functools import cache
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @cache
        def dfs(cur_amount):
            if cur_amount == 0:
                return 0
            ans = float('inf')
            for coin in coins:
                if cur_amount - coin >= 0:
                    ans = min(ans, dfs(cur_amount - coin) + 1)
            return ans

        ans = dfs(amount)
        if ans == float('inf'):
            return -1

        return ans


sol = Solution()
print(sol.coinChange([195,265,404,396], 3239))